pkg/volume/rbd/rbd.go

```diff
@@ -486,16 +521,95 @@ func (c *rbdUnmounter) TearDownAt(dir string) error {
+func getVolumeSourceMonitors(spec *volume.Spec) ([]string, error) {
+	if spec.Volume != nil && spec.Volume.RBD != nil {
+		mon := spec.Volume.RBD.CephMonitors
+		return mon, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.RBD != nil {
+		mon := spec.PersistentVolume.Spec.RBD.CephMonitors
+		return mon, nil
+	}
+
+	return nil, fmt.Errorf("Spec does not reference a RBD volume type")
+}
+
+func getVolumeSourceImage(spec *volume.Spec) (string, error) {
 	if spec.Volume != nil && spec.Volume.RBD != nil {
-		return spec.Volume.RBD, spec.Volume.RBD.ReadOnly, nil
+		image := spec.Volume.RBD.RBDImage
+		return image, nil
 	} else if spec.PersistentVolume != nil &&
 		spec.PersistentVolume.Spec.RBD != nil {
-		return spec.PersistentVolume.Spec.RBD, spec.ReadOnly, nil
+		image := spec.PersistentVolume.Spec.RBD.RBDImage
+		return image, nil
 	}
 
-	return nil, false, fmt.Errorf("Spec does not reference a RBD volume type")
+	return "", fmt.Errorf("Spec does not reference a RBD volume type")
+}
+
+func getVolumeSourceFSType(spec *volume.Spec) (string, error) {
+	if spec.Volume != nil && spec.Volume.RBD != nil {
+		fsType := spec.Volume.RBD.FSType
+		return fsType, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.RBD != nil {
+		fsType := spec.PersistentVolume.Spec.RBD.FSType
+		return fsType, nil
+	}
+
+	return "", fmt.Errorf("Spec does not reference a RBD volume type")
+}
+
+func getVolumeSourcePool(spec *volume.Spec) (string, error) {
+	if spec.Volume != nil && spec.Volume.RBD != nil {
+		pool := spec.Volume.RBD.RBDPool
+		return pool, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.RBD != nil {
+		pool := spec.PersistentVolume.Spec.RBD.RBDPool
+		return pool, nil
+	}
+
+	return "", fmt.Errorf("Spec does not reference a RBD volume type")
+}
+
+func getVolumeSourceUser(spec *volume.Spec) (string, error) {
+	if spec.Volume != nil && spec.Volume.RBD != nil {
+		user := spec.Volume.RBD.RadosUser
+		return user, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.RBD != nil {
+		user := spec.PersistentVolume.Spec.RBD.RadosUser
+		return user, nil
+	}
+
+	return "", fmt.Errorf("Spec does not reference a RBD volume type")
+}
+
+func getVolumeSourceKeyRing(spec *volume.Spec) (string, error) {
+	if spec.Volume != nil && spec.Volume.RBD != nil {
+		keyRing := spec.Volume.RBD.Keyring
+		return keyRing, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.RBD != nil {
+		keyRing := spec.PersistentVolume.Spec.RBD.Keyring
+		return keyRing, nil
+	}
+
+	return "", fmt.Errorf("Spec does not reference a RBD volume type")
+}
+
+func getVolumeSourceReadOnly(spec *volume.Spec) (bool, error) {
+	if spec.Volume != nil && spec.Volume.RBD != nil {
+		readOnly := spec.Volume.RBD.ReadOnly
+		return readOnly, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.RBD != nil {
+		readOnly := spec.ReadOnly
```

Comment:
Comment here that this is intentional would be good (also worth noting in the API doc for this field that it is ignored in favor of the PV readonly setting would be good)

Reply:
done